src/software/new_geom/util/distance.h

```diff
@@ -1,14 +1,76 @@
 #pragma once
 
+#include <limits>
+
+#include "software/geom/util.h"
 #include "software/new_geom/line.h"
 #include "software/new_geom/point.h"
+#include "software/new_geom/polygon.h"
+#include "software/new_geom/segment.h"
 
 /**
  * Finds the shortest distance between a Line and a Point
  *
- * @param first the Line
- * @param second the Point
+ * @param first
+ * @param second
  * @return the shortest distance between first and second
  */
 double distance(const Line &first, const Point &second);
 double distance(const Point &first, const Line &second);
+
+/**
+ * Finds the shortest distance between two Points
+ *
+ * @param first
+ * @param second
+ * @return the shortest distance between first and second
+ */
+double distance(const Point &first, const Point &second);
+
+/**
+ * Finds the shortest distance between two Segments
+ *
+ * @param first
+ * @param second
+ * @return the shortest distance between first and second
+ */
+double distance(const Segment &first, const Segment &second);
+
+/**
+ * Finds the shortest distance between a Point and a Segment
+ *
+ * @param first
+ * @param second
+ * @return the shortest distance between first and second
+ */
+double distance(const Point &first, const Segment &second);
+double distance(const Segment &first, const Point &second);
+
+/**
+ * Finds the shortest distance between a Point and a Polygon
+ *
+ * @param first
+ * @param second
+ * @return the shortest distance between first and second
+ */
+double distance(const Point &first, const Polygon &second);
+double distance(const Polygon &first, const Point &second);
+
+/**
+ * Finds the squared shortest distance between a Point and a Segment
+ *
+ * @param first
+ * @param second
+ * @return the squared shortest distance between first and second
+ */
+double distanceSquared(const Point &first, const Segment &second);
```

Comment:
More general comment, do we even really need `distanceSquared` functions? The reason there was so much support for it originally was that it's faster for comparing stuff (especially pythagoras) because it saves a `sqrt` operation.

For 99% of cases I think this is overkill, and since we don't have `distanceSquared` versions of all current `distance` functions, perhaps we should just be consistent and only have `distance` functions?

@garethellis0 @jonathanlew @jsn0716 for your thoughts

Reply:
I think making it consistent would be good. The usages for `distanceSquared` functions are pretty low.

Reply:
Seems like what you're trying to get at is we're effectively implementing `==` with distanceSquared. So instead we should just implement `==` (not sure how exactly) and optimize that by not taking a sqrt

Reply:
Sorry, not sure if I'm misreading this, but how are we effectively implementing `==`? `distanceSquared` is used for a fair bit more then just equality checks I would imagine?